File: packages/lessonplan/tts.go
package lessonplan

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"

	"t2t.dev/lessonplan/audio"
	"t2t.dev/lessonplan/domain"
)

// generateTTS calls OpenAI TTS for each segment, saves MP3 files, and measures durations.
func (g *Generator) generateTTS(ctx context.Context, segments []domain.NarrationSegment, lessonID string) ([]domain.AudioSegment, error) {
	audioDir := filepath.Join(g.outputDir, lessonID, "audio")
	if err := os.MkdirAll(audioDir, 0o755); err != nil {
		return nil, fmt.Errorf("create audio dir: %w", err)
	}

	var result []domain.AudioSegment
	for _, seg := range segments {
		mp3Data, err := g.callTTS(ctx, seg.Text)
		if err != nil {
			return nil, fmt.Errorf("segment %s: %w", seg.SegmentID, err)
		}

		filePath := filepath.Join(audioDir, seg.SegmentID+".mp3")
		if err := os.WriteFile(filePath, mp3Data, 0o644); err != nil {
			return nil, fmt.Errorf("write %s: %w", filePath, err)
		}

		duration, err := audio.MeasureDuration(filePath)
		if err != nil {
			// fallback: estimate from file size at 128 kbps
			duration = round3(float64(len(mp3Data)) * 8.0 / 128_000.0)
		}

		result = append(result, domain.AudioSegment{
			SegmentID:   seg.SegmentID,
			Text:        seg.Text,
			AudioFile:   "audio/" + seg.SegmentID + ".mp3",
			DurationSec: duration,
			Start:       0, // assigned by timeline step
		})
	}
	return result, nil
}

func (g *Generator) callTTS(ctx context.Context, text string) ([]byte, error) {
	payload, _ := json.Marshal(map[string]string{
		"model":           "tts-1",
		"input":           text,
		"voice":           g.voice,
		"response_format": "mp3",
	})

	req, err := http.NewRequestWithContext(ctx, http.MethodPost,
		"https://api.openai.com/v1/audio/speech", bytes.NewReader(payload))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Authorization", "Bearer "+g.apiKey)
	req.Header.Set("Content-Type", "application/json")

	resp, err := g.httpClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("status %d: %s", resp.StatusCode, body)
	}
	return io.ReadAll(resp.Body)
}

// concatenateAudio joins individual MP3 segments into a single full.mp3.
func concatenateAudio(segments []domain.AudioSegment, lessonID, outputDir string) (string, error) {
	fullPath := filepath.Join(outputDir, lessonID, "audio", "full.mp3")
	out, err := os.Create(fullPath)
	if err != nil {
		return "", fmt.Errorf("create full.mp3: %w", err)
	}
	defer out.Close()

	for _, seg := range segments {
		part, err := os.ReadFile(filepath.Join(outputDir, lessonID, seg.AudioFile))
		if err != nil {
			return "", fmt.Errorf("read %s: %w", seg.AudioFile, err)
		}
		if _, err := out.Write(part); err != nil {
			return "", fmt.Errorf("write full.mp3: %w", err)
		}
	}
	return "audio/full.mp3", nil
}

File: apps/lesson-plan-generator/app/steps/tts.py
"""Steps 3 & 4 – Generate TTS audio and measure segment durations.

Uses OpenAI TTS API (tts-1 model) to produce per-segment MP3 files, then
reads duration via mutagen.  Falls back to a bitrate estimate if mutagen fails.

The full audio track is assembled by concatenating raw MP3 bytes – no ffmpeg
dependency required.  For production-quality gapless audio consider pydub/ffmpeg.
"""

from pathlib import Path

from openai import AsyncOpenAI

from app.models import AudioSegment, NarrationSegment


async def generate_tts(
    segments: list[NarrationSegment],
    lesson_id: str,
    output_dir: Path,
    client: AsyncOpenAI,
    voice: str = "alloy",
) -> list[AudioSegment]:
    """Step 3: call OpenAI TTS for each segment and save MP3 files.
    Step 4: measure each file's duration.
    Returns AudioSegments with duration_sec set; start will be filled by timeline step.
    """
    audio_dir = output_dir / lesson_id / "audio"
    audio_dir.mkdir(parents=True, exist_ok=True)

    result: list[AudioSegment] = []
    for seg in segments:
        file_path = audio_dir / f"{seg.segment_id}.mp3"

        response = await client.audio.speech.create(
            model="tts-1",
            voice=voice,
            input=seg.text,
            response_format="mp3",
        )
        file_path.write_bytes(response.content)

        duration = _measure_duration(file_path)
        result.append(
            AudioSegment(
                segment_id=seg.segment_id,
                text=seg.text,
                audio_file=f"audio/{seg.segment_id}.mp3",
                duration_sec=duration,
                start=0.0,  # assigned by the timeline step
            )
        )

    return result


def concatenate_audio(
    segments: list[AudioSegment],
    lesson_id: str,
    output_dir: Path,
) -> str:
    """Concatenate individual MP3 files into a single full.mp3.

    Returns the relative path string stored in the video plan.
    """
    audio_dir = output_dir / lesson_id / "audio"
    full_path = audio_dir / "full.mp3"

    with open(full_path, "wb") as out:
        for seg in segments:
            part = output_dir / lesson_id / seg.audio_file
            out.write(part.read_bytes())

    return "audio/full.mp3"


def _measure_duration(file_path: Path) -> float:
    """Return MP3 duration in seconds.

    Primary: mutagen reads VBR/CBR headers accurately.
    Fallback: estimate from file size at 128 kbps.
    """
    try:
        from mutagen.mp3 import MP3

        info = MP3(str(file_path)).info
        return round(info.length, 3)
    except Exception:
        size_bytes = file_path.stat().st_size
        return round(size_bytes * 8 / 128_000, 3)
